tudatBundle/pagmo2/include/pagmo/algorithms/GACO.hpp

```diff
@@ -0,0 +1,1359 @@
+ * mixed integer non-linear programming. Computers & Operations Research.
+ */
+class g_aco
+{
+public:
+    /// Single entry of the log (gen, fevals, ideal_point)
+    typedef std::tuple<unsigned, double, double, double, double, double, double, double, double, double> log_line_type;
+    /// The log
+    typedef std::vector<log_line_type> log_type;
+
+    /// Constructors
+    /**
+    * Empty constructor
+    * */
+
+    // g_aco( ){ }
+
+    /**
+    * Constructs the ACO user defined algorithm for multi-objective optimization.
+    *
+    * @param[in] gen Generations: number of generations to evolve.
+    * @param[in] acc Accuracy parameter: for inequality and equality constraints .
+    * @param[in] fstop Objective stopping criterion: when the objective value reaches this value, the algorithm is stopped [for multi-objective, this applies to the first obj. only].
+    * @param[in] impstop Improvement stopping criterion: if a positive integer is assigned here, the algorithm will count the runs without improvements, if this number will exceed IMPSTOP value, the algorithm will be stopped.
+    * @param[in] evalstop Evaluation stopping criterion: same as previous one, but with function evaluations
+    * @param[in] focus Focus parameter: this parameter makes the search for the optimum greedier and more focused on local improvements (the higher the greedier). If the value is very high, the search is more focused around the current best solutions
+    * @param[in] ker Kernel: number of solutions stored in the solution archive
+    * @param[in] oracle Oracle parameter: this is the oracle parameter used in the penalty method
+    * @param[in] paretomax Max number of non-dominated solutions: this regulates the max number of Pareto points to be stored
+    * @param[in] epsilon Pareto precision: the smaller this parameter, the higher the chances to introduce a new solution in the Pareto front
+    * @param seed seed used by the internal random number generator (default is random)
+    * @throws std::invalid_argument if \p acc is not \f$ \in [0,1[\f$, \p impstop is not a
+    * positive integer, \p evalstop is not a positive integer, \p focus is not \f$ \in [0,1[\f$, \p ants is not a positive integer,
+    * \p ker is not a positive integer, \p oracle is not positive, \p paretomax is not a positive integer, \p epsilon is not \f$ \in [0,1[\f$
+    */
+    g_aco(unsigned gen = 100u, double acc = 0.95, double  fstop = 0.0000001, int impstop = 100, int evalstop = 100,
+          double focus = 0,  unsigned ker = 250, double oracle=10000, unsigned paretomax = 10,
+            double epsilon = 0.9, unsigned seed = pagmo::random_device::next())
+        : m_gen(gen), m_acc(acc), m_fstop(fstop), m_impstop(impstop), m_evalstop(evalstop), m_focus(focus),
+          m_ker(ker), m_oracle(oracle), m_paretomax(paretomax), m_epsilon(epsilon), m_e(seed), m_seed(seed), m_verbosity(0u),
+          m_log(), res()
```

Comment:
@darioizzo  @bluescarni 
These last days I debugged the algorithm, which was not working properly and had several problems. 

I ran the Rosenbrock function with:

    pagmo::problem prob{rosenbrock{10}};

    // Algorithm (setting generations to 100)
    pagmo::algorithm algo{ g_aco {30} };

    // Set the algo to log something at each iteration
    algo.set_verbosity(1);

    // Problem
    //pagmo::problem prob{rosenbrock{10}};

    // Population
    pagmo::population pop{prob, 1000};

It turns out that it reaches a value of 9.37276 for the best found fitness. I compared it with de1220 and it turns out that this second one reaches 67.8776. So I guess it is quite reasonable as result. I changed the parameters because of some characteristics of the GACO algorithm, which prefers a bigger number of population and smaller of generation, rather than the opposite.

What do you suggest as next steps for the testing?

(PS. Out of curiosity, I also ran the Himmelblau and I managed to converge to something like 1e-5, with not many population and generations).

Reply:
@bluescarni 
I think I lack a clear step-by-step procedure for testing a new algorithm. Do you recommend a specific procedure to be followed?

Reply:
@Sceki, to avoid cluttering this too much, please use gitter for questions that are not on specific commit lines and are general on coding, testing programming etc.

Reply:
@Sceki because we are talking about metaheuristics, it is difficult to rigorously "prove" that the implementation is correct. So practically, in this context, the algorithm is considered to be "working" if it produces a population that is not worse (for some definition of worse...) after an evolution.

You must also ensure that all the codepaths are executed at least once (including error handling). That is, each line of the algorithm code should be exercised in the test. You can take a look at the other algorithms' tests if you need inspiration, if you have any issue please come over to the gitter channel and let's have a chat there.

Reply:
Thank you. I will have a look and contact you in Ghitter in case I will have problems.